@mnencia it is not working for me. If I run:

```sh
task e2e:test:full TARGET=wal2json
```

I used this basic Dockerfile:

```dockerfile
# SPDX-FileCopyrightText: Copyright © contributors to CloudNativePG, established as CloudNativePG a Series of LF Projects, LLC.
# SPDX-License-Identifier: Apache-2.0

ARG BASE=ghcr.io/cloudnative-pg/postgresql:18-minimal-trixie
FROM $BASE AS builder

ARG PG_MAJOR
ARG EXT_VERSION

USER 0

RUN apt-get update && apt-get install -y --no-install-recommends \
  "postgresql-${PG_MAJOR}-wal2json=${EXT_VERSION}"

FROM scratch
ARG PG_MAJOR

# Licenses
# COPY --from=builder /usr/share/doc/postgresql-${PG_MAJOR}-wal2json/copyright /licenses/postgresql-${PG_MAJOR}-wal2json/

# Libraries
# COPY --from=builder /usr/lib/postgresql/${PG_MAJOR}/lib/wal2json* /lib/

USER 65532:65532
```

And this metadata:

```hcl
# SPDX-FileCopyrightText: Copyright © contributors to CloudNativePG, established as CloudNativePG a Series of LF Projects, LLC.
# SPDX-License-Identifier: Apache-2.0
metadata = {
  name                     = "wal2json"
  sql_name                 = "wal2json"
  image_name               = "wal2json"
  shared_preload_libraries = []
  extension_control_path   = []
  dynamic_library_path     = []
  ld_library_path          = []
  auto_update_os_libs      = false
  required_extensions      = []
  create_extension         = false

  versions = {
    trixie = {
        // renovate: suite=trixie-pgdg depName=postgresql-18-wal2json
        "18" = "2.6-3.pgdg13+1"
    }
    bookworm = {
        // renovate: suite=bookworm-pgdg depName=postgresql-18-wal2json
        "18" = "2.6-3.pgdg12+1"
    }
  }
}
```

I get this result:

```console
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | ASSERT    | ERROR | postgresql.cnpg.io/v1/Database @ chainsaw-innocent-toucan/wal2json-app
[e2e:test-wal2json]         === ERROR
[e2e:test-wal2json]         --------------------------------------------------------------------
[e2e:test-wal2json]         postgresql.cnpg.io/v1/Database/chainsaw-innocent-toucan/wal2json-app
[e2e:test-wal2json]         --------------------------------------------------------------------
[e2e:test-wal2json]         * status.extensions: Required value: field not found in the input object
[e2e:test-wal2json]
[e2e:test-wal2json]         --- expected
[e2e:test-wal2json]         +++ actual
[e2e:test-wal2json]         @@ -5,6 +5,5 @@
[e2e:test-wal2json]            namespace: chainsaw-innocent-toucan
[e2e:test-wal2json]          status:
[e2e:test-wal2json]            applied: true
[e2e:test-wal2json]         -  extensions: null
[e2e:test-wal2json]            observedGeneration: 1
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | TRY       | END   |
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | CLEANUP   | BEGIN |
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | DELETE    | OK    | postgresql.cnpg.io/v1/Database @ chainsaw-innocent-toucan/wal2json-app
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | DELETE    | OK    | postgresql.cnpg.io/v1/Cluster @ chainsaw-innocent-toucan/wal2json
[e2e:test-wal2json]     | 03:48:54 | verify-extension | Create a Cluster with the extension | CLEANUP   | END   |
[e2e:test-wal2json]     | 03:48:54 | verify-extension | @chainsaw                           | CLEANUP   | BEGIN |
[e2e:test-wal2json]     | 03:48:54 | verify-extension | @chainsaw                           | DELETE    | OK    | v1/Namespace @ chainsaw-innocent-toucan
[e2e:test-wal2json]     | 03:49:04 | verify-extension | @chainsaw                           | CLEANUP   | END   |
[e2e:test-wal2json] --- FAIL: chainsaw (0.00s)
[e2e:test-wal2json]     --- FAIL: chainsaw/verify-extension (209.20s)
[e2e:test-wal2json] FAIL
[e2e:test-wal2json] Tests Summary...
[e2e:test-wal2json] - Passed  tests 0
[e2e:test-wal2json] - Failed  tests 1
[e2e:test-wal2json] - Skipped tests 0
[e2e:test-wal2json] Done with failures.
[e2e:test-wal2json] Stderr:
[e2e:test-wal2json] Error: some tests failed
task: [e2e:test:full] rm -f "./kubeconfig"
task: [e2e:test:full] rm -f "wal2json/values.yaml"
task: [e2e:test:full] rm -f "wal2json/bake-metadata.json"
task: Failed to run task "e2e:test:full": task: Failed to run task "e2e:test": exit status 1
...
```

--- a/dagger/maintenance/dagger.json
+++ b/dagger/maintenance/dagger.json
@@ -1,6 +1,6 @@
 {
   "name": "maintenance",
-  "engineVersion": "v0.19.7",
+  "engineVersion": "v0.19.10",
   "sdk": {
     "source": "go"
   }

--- a/dagger/maintenance/go.mod
+++ b/dagger/maintenance/go.mod
@@ -88,7 +88,7 @@ require (
 	github.com/moby/term v0.5.2 // indirect
 	github.com/morikuni/aec v1.0.0 // indirect
 	github.com/opencontainers/go-digest v1.0.0 // indirect
-	github.com/opencontainers/image-spec v1.1.1 // indirect
+	github.com/opencontainers/image-spec v1.1.1
 	github.com/pelletier/go-toml v1.9.5 // indirect
 	github.com/pkg/errors v0.9.1 // indirect
 	github.com/planetscale/vtprotobuf v0.6.1-0.20240319094008-0393e58bdf10 // indirect
@@ -130,7 +130,7 @@ require (
 	golang.org/x/net v0.44.0 // indirect
 	golang.org/x/sync v0.17.0
 	golang.org/x/sys v0.37.0 // indirect
-	golang.org/x/text v0.29.0 // indirect
+	golang.org/x/text v0.29.0
 	golang.org/x/time v0.14.0 // indirect
 	golang.org/x/tools v0.37.0 // indirect
 	google.golang.org/genproto/googleapis/api v0.0.0-20250825161204-c5933d9347a5 // indirect

--- a/dagger/maintenance/main.go
+++ b/dagger/maintenance/main.go
@@ -185,28 +185,20 @@ func (m *Maintenance) GenerateTestingValues(
 		return nil, err
 	}
 
-	var extensionsToCreate []map[string]string
-	var expectedExtensions []map[string]any
-	if metadata.CreateExtension {
-		extensionsToCreate = []map[string]string{
-			{"name": metadata.SQLName, "version": version},
-		}
-		expectedExtensions = []map[string]any{
-			{"applied": true, "name": metadata.SQLName},
-		}
+	databaseConfig, err := generateDatabaseConfig(ctx, source, extensions)
+	if err != nil {
+		return nil, err
 	}
 
 	// Build values.yaml content
-	values := map[string]any{
-		"name":                     metadata.Name,
-		"sql_name":                 metadata.SQLName,
-		"shared_preload_libraries": metadata.SharedPreloadLibraries,
-		"pg_image":                 pgImage,
-		"version":                  version,
-		"extensions":               extensions,
-		"create_extension":         metadata.CreateExtension,
-		"extensions_to_create":     extensionsToCreate,
-		"expected_extensions":      expectedExtensions,
+	values := TestingValues{
+		Name:                   metadata.Name,
+		SQLName:                metadata.SQLName,
+		SharedPreloadLibraries: metadata.SharedPreloadLibraries,
+		PgImage:                pgImage,
+		Version:                version,
+		Extensions:             extensions,
+		DatabaseConfig:         databaseConfig,
 	}
 	valuesYaml, err := yaml.Marshal(values)
 	if err != nil {

--- a/dagger/maintenance/main.go
+++ b/dagger/maintenance/main.go
@@ -185,6 +185,17 @@ func (m *Maintenance) GenerateTestingValues(
 		return nil, err
 	}
 
+	var extensionsToCreate []map[string]string
+	var expectedExtensions []map[string]any
+	if metadata.CreateExtension {
+		extensionsToCreate = []map[string]string{
+			{"name": metadata.SQLName, "version": version},
+		}
+		expectedExtensions = []map[string]any{
+			{"applied": true, "name": metadata.SQLName},
+		}
+	}
+
 	// Build values.yaml content
 	values := map[string]any{
 		"name":                     metadata.Name,
@@ -193,6 +204,9 @@ func (m *Maintenance) GenerateTestingValues(
 		"pg_image":                 pgImage,
 		"version":                  version,
 		"extensions":               extensions,
+		"create_extension":         metadata.CreateExtension,
+		"extensions_to_create":     extensionsToCreate,
+		"expected_extensions":      expectedExtensions,
 	}
 	valuesYaml, err := yaml.Marshal(values)
 	if err != nil {

--- a/dagger/maintenance/parse.go
+++ b/dagger/maintenance/parse.go
@@ -30,6 +30,7 @@ type extensionMetadata struct {
 	LdLibraryPath          []string   `hcl:"ld_library_path" cty:"ld_library_path"`
 	AutoUpdateOsLibs       bool       `hcl:"auto_update_os_libs" cty:"auto_update_os_libs"`
 	RequiredExtensions     []string   `hcl:"required_extensions" cty:"required_extensions"`
+	CreateExtension        bool       `hcl:"create_extension" cty:"create_extension"`
 	Versions               versionMap `hcl:"versions" cty:"versions"`
 	Remain                 hcl.Body   `hcl:",remain"`
 }

--- a/dagger/maintenance/testingvalues.go
+++ b/dagger/maintenance/testingvalues.go
@@ -7,8 +7,34 @@ import (
 	"dagger/maintenance/internal/dagger"
 )
 
-func generateTestingValuesExtensions(ctx context.Context, source *dagger.Directory, metadata *extensionMetadata, extensionImage string) ([]map[string]any, error) {
-	var out []map[string]any
+type ExtensionSpec struct {
+	Ensure  string `yaml:"ensure"`
+	Name    string `yaml:"name"`
+	Version string `yaml:"version"`
+}
+
+type ExpectedStatus struct {
+	Applied bool   `yaml:"applied"`
+	Name    string `yaml:"name"`
+}
+
+type DatabaseConfig struct {
+	ExtensionsSpec []ExtensionSpec  `yaml:"extensions_spec"`
+	ExpectedStatus []ExpectedStatus `yaml:"expected_status"`
+}
+
+type TestingValues struct {
+	Name                   string                    `yaml:"name"`
+	SQLName                string                    `yaml:"sql_name"`
+	SharedPreloadLibraries []string                  `yaml:"shared_preload_libraries"`
+	PgImage                string                    `yaml:"pg_image"`
+	Version                string                    `yaml:"version"`
+	Extensions             []*ExtensionConfiguration `yaml:"extensions"`
+	DatabaseConfig         *DatabaseConfig           `yaml:"database_config"`
+}
+
+func generateTestingValuesExtensions(ctx context.Context, source *dagger.Directory, metadata *extensionMetadata, extensionImage string) ([]*ExtensionConfiguration, error) {
+	var out []*ExtensionConfiguration
 	configuration, err := generateExtensionConfiguration(metadata, extensionImage)
 	if err != nil {
 		return nil, err
@@ -24,21 +50,21 @@ func generateTestingValuesExtensions(ctx context.Context, source *dagger.Directo
 			return nil, fmt.Errorf("required dependency %q not found", dep)
 		}
 
-		depMetadata, parseErr := parseExtensionMetadata(ctx, source.Directory(dep))
-		if parseErr != nil {
-			return nil, fmt.Errorf("failed to parse dependency metadata %q: %w", dep, parseErr)
+		depMetadata, err := parseExtensionMetadata(ctx, source.Directory(dep))
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse dependency metadata %q: %w", dep, err)
 		}
-		depsConfiguration, extErr := generateExtensionConfiguration(depMetadata, "")
-		if extErr != nil {
-			return nil, extErr
+		depConfiguration, err := generateExtensionConfiguration(depMetadata, "")
+		if err != nil {
+			return nil, err
 		}
-		out = append(out, depsConfiguration)
+		out = append(out, depConfiguration)
 	}
 
 	return out, nil
 }
 
-func generateExtensionConfiguration(metadata *extensionMetadata, extensionImage string) (map[string]any, error) {
+func generateExtensionConfiguration(metadata *extensionMetadata, extensionImage string) (*ExtensionConfiguration, error) {
 	targetExtensionImage := extensionImage
 	if targetExtensionImage == "" {
 		var err error
@@ -48,13 +74,56 @@ func generateExtensionConfiguration(metadata *extensionMetadata, extensionImage 
 		}
 	}
 
-	return map[string]any{
-		"name": metadata.Name,
-		"image": map[string]string{
-			"reference": targetExtensionImage,
+	return &ExtensionConfiguration{
+		Name: metadata.Name,
+		ImageVolumeSource: ImageVolumeSource{
+			Reference: targetExtensionImage,
 		},
-		"extension_control_path": metadata.ExtensionControlPath,
-		"dynamic_library_path":   metadata.DynamicLibraryPath,
-		"ld_library_path":        metadata.LdLibraryPath,
+		ExtensionControlPath: metadata.ExtensionControlPath,
+		DynamicLibraryPath:   metadata.DynamicLibraryPath,
+		LdLibraryPath:        metadata.LdLibraryPath,
 	}, nil
+}
+
+func generateDatabaseConfig(ctx context.Context, source *dagger.Directory, extensionsConfig []*ExtensionConfiguration) (*DatabaseConfig, error) {
+	var databaseConfig DatabaseConfig
+	for _, extension := range extensionsConfig {
+		extMetadata, err := parseExtensionMetadata(ctx, source.Directory(extension.Name))
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse dependency metadata %q: %w", extension.Name, err)
+		}
+
+		extAnnotations, err := getImageAnnotations(extension.ImageVolumeSource.Reference)
+		if err != nil {
+			return nil, err
+		}
+
+		extVersion := extAnnotations["org.opencontainers.image.version"]
+		if extVersion == "" {
+			return nil, fmt.Errorf(
+				"extension image %s doesn't have an 'org.opencontainers.image.version' annotation",
+				extension.ImageVolumeSource.Reference)
+		}
+
+		ensureOption := "absent"
+		if extMetadata.CreateExtension {
+			ensureOption = "present"
+		}
+
+		databaseConfig.ExtensionsSpec = append(databaseConfig.ExtensionsSpec,
+			ExtensionSpec{
+				Ensure:  ensureOption,
+				Name:    extMetadata.SQLName,
+				Version: extVersion,
+			},
+		)
+		databaseConfig.ExpectedStatus = append(databaseConfig.ExpectedStatus,
+			ExpectedStatus{
+				Name:    extMetadata.SQLName,
+				Applied: true,
+			},
+		)
+	}
+
+	return &databaseConfig, nil
 }

--- a/pgaudit/metadata.hcl
+++ b/pgaudit/metadata.hcl
@@ -8,6 +8,7 @@ metadata = {
   ld_library_path          = []
   auto_update_os_libs      = false
   required_extensions      = []
+  create_extension         = true
 
   versions = {
     bookworm = {

--- a/pgvector/metadata.hcl
+++ b/pgvector/metadata.hcl
@@ -8,6 +8,7 @@ metadata = {
   ld_library_path          = []
   auto_update_os_libs      = false
   required_extensions      = []
+  create_extension         = true
 
   versions = {
     bookworm = {

--- a/postgis/metadata.hcl
+++ b/postgis/metadata.hcl
@@ -8,6 +8,7 @@ metadata = {
   ld_library_path          = ["system"]
   auto_update_os_libs      = true
   required_extensions      = []
+  create_extension         = true
 
   versions = {
     bookworm = {

--- a/templates/metadata.hcl.tmpl
+++ b/templates/metadata.hcl.tmpl
@@ -47,6 +47,7 @@ metadata = {
   # `required_extensions`: must contain the name(s) of the sibling
   # folders in this repository that contain a required extension.
   required_extensions      = []
+  create_extension         = true
 
   versions = {
     {{- range $distro := .Distros}}

--- a/templates/metadata.hcl.tmpl
+++ b/templates/metadata.hcl.tmpl
@@ -47,6 +47,12 @@ metadata = {
   # `required_extensions`: must contain the name(s) of the sibling
   # folders in this repository that contain a required extension.
   required_extensions      = []
+
+  # TODO: Remove this comment block after customizing the file.
+  # `create_extension`: if set to `true` (default), the test suite will
+  # automatically run `CREATE EXTENSION` for this project during E2E tests.
+  # Set to `false` if the image only provides libraries or tools without
+  # a formal Postgres extension object.
   create_extension         = true
 
   versions = {

--- a/test/check-extension.yaml
+++ b/test/check-extension.yaml
@@ -13,6 +13,8 @@ spec:
             value: ($values.sql_name)
           - name: EXT_VERSION
             value: ($values.version)
+          - name: CREATE_EXTENSION
+            value: (to_string($values.create_extension))
           - name: DB_URI
             valueFrom:
               secretKeyRef:
@@ -23,6 +25,10 @@ spec:
         args:
          - |
            set -e
+           if [ "$CREATE_EXTENSION" != "true" ]; then
+             echo "Skipping extension check (create_extension=false)"
+             exit 0
+           fi
            DB_URI=$(echo $DB_URI | sed "s|/\*|/|")
            test "$(psql "$DB_URI" -tAc "SELECT EXISTS (SELECT FROM pg_catalog.pg_extension WHERE extname = '${EXT_SQL_NAME}' AND extversion = '${EXT_VERSION}')" -q)" = "t"
            echo "Extension '${EXT_SQL_NAME} v${EXT_VERSION}' is installed!"

--- a/test/database-assert.yaml
+++ b/test/database-assert.yaml
@@ -4,3 +4,5 @@ metadata:
   name: (join('-', [$values.name, 'app']))
 status:
   applied: true
+  extensions: ($values.database_config.expected_status)
+  observedGeneration: 1

--- a/test/database-assert.yaml
+++ b/test/database-assert.yaml
@@ -4,7 +4,3 @@ metadata:
   name: (join('-', [$values.name, 'app']))
 status:
   applied: true
-  extensions:
-  - applied: true
-    name: ($values.sql_name)
-  observedGeneration: 1

--- a/test/database.yaml
+++ b/test/database.yaml
@@ -7,4 +7,4 @@ spec:
   owner: app
   cluster:
     name: ($values.name)
-  extensions: ($values.extensions_to_create)
+  extensions: ($values.database_config.extensions_spec)

--- a/test/database.yaml
+++ b/test/database.yaml
@@ -7,6 +7,4 @@ spec:
   owner: app
   cluster:
     name: ($values.name)
-  extensions:
-  - name: ($values.sql_name)
-    version: ($values.version)
+  extensions: ($values.extensions_to_create)

